File: client/node/conformance/compare.ts
// Comparing a normalized response against a scenario's expectation, by the
// rules in conformance/README.md.

import { Integer, type Normalized } from "./normalize.js";
import { byCodeUnit } from "./order.js";
import { Literal, type Expect } from "./scenarios.js";

/**
 * The relative difference two Real values may have and still be the same value.
 * It applies to Reals alone: an integral field is compared by its digits.
 */
export const TOLERANCE = 1e-9;

/** Compares a normalized response against one expectation, one message per mismatch. */
export function check(expect: Expect, actual: Normalized): string[] {
  const failures: string[] = [];
  if (expect.response !== undefined) {
    failures.push(...match("", expect.response, actual));
  }
  for (const path of [...(expect.non_empty ?? [])].sort(byCodeUnit)) {
    const found = lookup(actual, path);
    if (!found.ok) {
      failures.push(`${path}: not set, want a value`);
    } else if (isDefault(found.value)) {
      failures.push(`${path}: empty, want a value`);
    }
  }
  for (const path of [...(expect.absent ?? [])].sort(byCodeUnit)) {
    const found = lookup(actual, path);
    if (found.ok && !isDefault(found.value)) {
      failures.push(`${path}: set to ${render(found.value)}, want it unset`);
    }
  }
  for (const path of sortedKeys(expect.contains)) {
    const want = expect.contains?.[path] ?? "";
    const found = lookup(actual, path);
    if (!found.ok) {
      failures.push(`${path}: not set, want it to contain ${quote(want)}`);
    } else if (typeof found.value !== "string") {
      failures.push(`${path}: ${render(found.value)} is not text, want it to contain ${quote(want)}`);
    } else if (!found.value.includes(want)) {
      failures.push(`${path}: ${quote(found.value)} does not contain ${quote(want)}`);
    }
  }
  for (const path of sortedKeys(expect.contains_all)) {
    failures.push(...containsAll(actual, path, expect.contains_all?.[path] ?? []));
  }
  for (const path of sortedKeys(expect.counts)) {
    const want = numberOf(expect.counts?.[path]);
    const got = count(actual, path);
    if (got === undefined) {
      failures.push(`${path}: not a list or map, want ${want} entries`);
    } else if (got !== want) {
      failures.push(`${path}: ${got} entries, want ${want}`);
    }
  }
  for (const path of sortedKeys(expect.min_counts)) {
    const want = numberOf(expect.min_counts?.[path]);
    const got = count(actual, path);
    if (got === undefined) {
      failures.push(`${path}: not a list or map, want at least ${want} entries`);
    } else if (got < want) {
      failures.push(`${path}: ${got} entries, want at least ${want}`);
    }
  }
  return failures;
}

/**
 * Checks that every wanted string is at `path`: a substring of the text there, or
 * a member of the values there. `*` takes one field of every entry of a list.
 */
function containsAll(actual: Normalized, path: string, wants: string[]): string[] {
  const found = lookup(actual, path);
  if (found.ok && typeof found.value === "string") {
    const text = found.value;
    return wants.filter((want) => !text.includes(want)).map((want) => `${path}: does not contain ${quote(want)}`);
  }
  const collected = values(actual, path);
  if (collected === undefined) {
    return [`${path}: neither text nor a list, want it to contain ${render(wants)}`];
  }
  return wants
    .filter((want) => !collected.includes(want))
    .map((want) => `${path}: ${render(collected)} does not contain ${quote(want)}`);
}

/** The values at a path, expanding `*` and a trailing list into its members. */
function values(tree: unknown, path: string): unknown[] | undefined {
  const found = walk(tree, path.split("."));
  if (found === undefined) {
    return undefined;
  }
  if (found.length === 1 && Array.isArray(found[0])) {
    return found[0] as unknown[];
  }
  return found;
}

function walk(current: unknown, segments: string[]): unknown[] | undefined {
  if (segments.length === 0) {
    return [current];
  }
  const [head, ...rest] = segments;
  if (head === "*") {
    let entries: unknown[];
    if (Array.isArray(current)) {
      entries = current;
    } else if (isObject(current)) {
      entries = sortedKeys(current).map((key) => current[key]);
    } else {
      return undefined;
    }
    const found: unknown[] = [];
    for (const entry of entries) {
      const reached = walk(entry, rest);
      if (reached === undefined) {
        return undefined;
      }
      found.push(...reached);
    }
    return found;
  }
  const step = lookup(current, head);
  if (!step.ok) {
    return undefined;
  }
  return walk(step.value, rest);
}

/** Compares an expected tree against the response's, reporting the paths that differ. */
function match(path: string, want: unknown, got: unknown): string[] {
  if (want instanceof Literal) {
    return matchNumber(path, want.text, got);
  }
  if (Array.isArray(want)) {
    if (!Array.isArray(got)) {
      return [`${at(path)}: ${render(got)}, want a list`];
    }
    if (got.length !== want.length) {
      return [`${at(path)}: ${got.length} entries, want ${want.length} (${render(got)})`];
    }
    return want.flatMap((item, index) => match(join(path, String(index)), item, got[index]));
  }
  if (isObject(want)) {
    if (!isObject(got)) {
      return [`${at(path)}: ${render(got)}, want an object`];
    }
    const failures: string[] = [];
    for (const key of sortedKeys(want)) {
      const child = join(path, key);
      if (!(key in got)) {
        // An unset field and a field left at its default are the same thing on
        // the wire, so a default expectation still matches.
        if (!isDefault(want[key])) {
          failures.push(`${child}: not set, want ${render(want[key])}`);
        }
        continue;
      }
      failures.push(...match(child, want[key], got[key]));
    }
    return failures;
  }
  if (typeof want === "number") {
    return matchNumber(path, String(want), got);
  }
  if (String(want) !== String(got)) {
    return [`${at(path)}: ${render(got)}, want ${render(want)}`];
  }
  return [];
}

/** Compares an expected number, as the scenario wrote it, against the response's. */
function matchNumber(path: string, want: string, got: unknown): string[] {
  if (got instanceof Integer) {
    const digits = integerLiteral(want);
    if (digits === undefined || digits !== got.digits) {
      return [`${at(path)}: ${got.digits}, want ${want}`];
    }
    return [];
  }
  if (typeof got === "number") {
    const number = Number(want);
    if (Number.isNaN(number) || !near(got, number)) {
      return [`${at(path)}: ${got}, want ${want}`];
    }
    return [];
  }
  return [`${at(path)}: ${render(got)}, want the number ${want}`];
}

/** An expected number as the digits of a whole number, also accepting "1500.0". */
function integerLiteral(text: string): string | undefined {
  if (/^[+-]?\d+$/.test(text)) {
    return BigInt(text).toString();
  }
  const number = Number(text);
  if (!Number.isFinite(number) || !Number.isInteger(number) || !Number.isSafeInteger(number)) {
    return undefined;
  }
  return BigInt(number).toString();
}

/** Whether two Real values are the same within the tolerance. */
function near(got: number, want: number): boolean {
  if (got === want) {
    return true;
  }
  const scale = Math.max(Math.abs(got), Math.abs(want));
  return Math.abs(got - want) <= TOLERANCE * scale;
}

/** Whether a value is what an unset field holds. */
export function isDefault(value: unknown): boolean {
  if (value instanceof Literal) {
    return Number(value.text) === 0;
  }
  if (value instanceof Integer) {
    return BigInt(value.digits) === 0n;
  }
  if (value === null || value === undefined) {
    return true;
  }
  switch (typeof value) {
    case "boolean":
      return !value;
    case "number":
      return value === 0;
    case "bigint":
      return value === 0n;
    case "string":
      return value === "";
    default:
      break;
  }
  if (Array.isArray(value)) {
    return value.length === 0;
  }
  if (isObject(value)) {
    return Object.keys(value).length === 0;
  }
  return false;
}

/** Walks a dotted path into a normalized response: fields, map keys, list indices. */
export function lookup(tree: unknown, path: string): { ok: true; value: unknown } | { ok: false } {
  let current = tree;
  for (const segment of path.split(".")) {
    if (Array.isArray(current)) {
      const index = Number(segment);
      if (!Number.isInteger(index) || index < 0 || index >= current.length) {
        return { ok: false };
      }
      current = current[index];
      continue;
    }
    if (isObject(current)) {
      if (!(segment in current)) {
        return { ok: false };
      }
      current = current[segment];
      continue;
    }
    return { ok: false };
  }
  return { ok: true, value: current };
}

/** The number of entries of the list or map at `path`. */
function count(tree: Normalized, path: string): number | undefined {
  const found = lookup(tree, path);
  if (!found.ok) {
    // An empty list or map is an unset field, so nothing there is zero entries.
    return 0;
  }
  if (Array.isArray(found.value)) {
    return found.value.length;
  }
  if (isObject(found.value)) {
    return Object.keys(found.value).length;
  }
  return undefined;
}

function numberOf(value: Literal | number | undefined): number {
  if (value instanceof Literal) {
    return Number(value.text);
  }
  return value ?? 0;
}

function at(path: string): string {
  return path === "" ? "response" : path;
}

function join(path: string, segment: string): string {
  return path === "" ? segment : `${path}.${segment}`;
}

function sortedKeys(value: Record<string, unknown> | undefined): string[] {
  return value === undefined ? [] : Object.keys(value).sort(byCodeUnit);
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value) && !(value instanceof Integer) && !(value instanceof Literal);
}

function quote(text: string): string {
  return JSON.stringify(text);
}

/** A value as a scenario would write it. */
export function render(value: unknown): string {
  if (typeof value === "string") {
    return quote(value);
  }
  if (value === null || value === undefined) {
    return "nothing";
  }
  if (value instanceof Integer) {
    return value.digits;
  }
  if (value instanceof Literal) {
    return value.text;
  }
  if (Array.isArray(value)) {
    return `[${value.map(render).join(", ")}]`;
  }
  if (isObject(value)) {
    return `{${sortedKeys(value)
      .map((key) => `${key}: ${render(value[key])}`)
      .join(", ")}}`;
  }
  return typeof value === "bigint" ? value.toString() : JSON.stringify(value);
}

File: client/rust/conformance/src/compare.rs
use std::collections::HashMap;

use serde_json::Value;

use crate::scenario::Expect;
use opensysml::Status;

pub fn status_matches(expected: Option<&str>, actual: Status) -> bool {
    Status::from_canonical_name(expected.unwrap_or("OK")) == Some(actual)
}

pub fn compare(expect: &Expect, actual: &Value) -> Vec<String> {
    let mut failures = Vec::new();
    if let Some(response) = &expect.response {
        compare_value(response, actual, "$", &mut failures);
    }
    for path in &expect.non_empty {
        check_non_empty(path, actual, &mut failures);
    }
    for path in &expect.absent {
        check_absent(path, actual, &mut failures);
    }
    for (path, needle) in &expect.contains {
        check_contains(path, needle, actual, &mut failures);
    }
    for (path, needles) in &expect.contains_all {
        check_contains_all(path, needles, actual, &mut failures);
    }
    for (path, wanted) in &expect.counts {
        check_count(path, *wanted, false, actual, &mut failures);
    }
    for (path, wanted) in &expect.min_counts {
        check_count(path, *wanted, true, actual, &mut failures);
    }
    failures
}

fn check_non_empty(path: &str, actual: &Value, failures: &mut Vec<String>) {
    match lookup(actual, path) {
        Some(value) if !is_default(value) => {}
        Some(_) => failures.push(format!("{path} is empty")),
        None => failures.push(format!("{path} is absent")),
    }
}

fn check_absent(path: &str, actual: &Value, failures: &mut Vec<String>) {
    if let Some(value) = lookup(actual, path) {
        if !is_default(value) {
            failures.push(format!("{path} is not absent/default: {value}"));
        }
    }
}

fn check_contains(path: &str, needle: &str, actual: &Value, failures: &mut Vec<String>) {
    match lookup(actual, path) {
        Some(Value::String(value)) if value.contains(needle) => {}
        Some(value) => failures.push(format!("{path}={value} does not contain {needle:?}")),
        None => failures.push(format!("{path} is absent")),
    }
}

fn check_contains_all(path: &str, needles: &[String], actual: &Value, failures: &mut Vec<String>) {
    if path.split('.').any(|segment| segment == "*") {
        check_members_at(path, needles, actual, failures);
        return;
    }
    match lookup(actual, path) {
        Some(Value::String(value)) => {
            for needle in needles {
                if !value.contains(needle) {
                    failures.push(format!("{path} does not contain {needle:?}"));
                }
            }
        }
        Some(Value::Array(values)) => {
            for needle in needles {
                if !values.iter().any(|value| value == needle) {
                    failures.push(format!("{path} does not contain member {needle:?}"));
                }
            }
        }
        Some(_) | None => check_members_at(path, needles, actual, failures),
    }
}

/// Every needle must be one of the strings the path collects across wildcards.
fn check_members_at(path: &str, needles: &[String], actual: &Value, failures: &mut Vec<String>) {
    let values = values_at(actual, path);
    if values.is_empty() {
        failures.push(format!("{path} is absent or not text/list"));
        return;
    }
    for needle in needles {
        if !values
            .iter()
            .any(|value| value.as_str().is_some_and(|text| text == needle))
        {
            failures.push(format!("{path} does not contain member {needle:?}"));
        }
    }
}

fn check_count(
    path: &str,
    wanted: usize,
    minimum: bool,
    actual: &Value,
    failures: &mut Vec<String>,
) {
    let Some(value) = lookup(actual, path) else {
        failures.push(format!("{path} is absent"));
        return;
    };
    let count = match value {
        Value::Array(items) => items.len(),
        Value::Object(items) => items.len(),
        _ => {
            failures.push(format!("{path} is not a list/map"));
            return;
        }
    };
    if (minimum && count < wanted) || (!minimum && count != wanted) {
        let relation = if minimum { "at least" } else { "exactly" };
        failures.push(format!(
            "{path} has {count} entries, want {relation} {wanted}"
        ));
    }
}

fn compare_value(expected: &Value, actual: &Value, path: &str, failures: &mut Vec<String>) {
    match (expected, actual) {
        (Value::Object(wanted), Value::Object(got)) => {
            for (key, value) in wanted {
                let Some(actual) = got.get(key) else {
                    failures.push(format!("{path}.{key} is absent"));
                    continue;
                };
                compare_value(value, actual, &format!("{path}.{key}"), failures);
            }
        }
        (Value::Array(wanted), Value::Array(got)) => {
            if wanted.len() != got.len() {
                failures.push(format!(
                    "{path} has {} entries, want {}",
                    got.len(),
                    wanted.len()
                ));
                return;
            }
            for (index, (wanted, got)) in wanted.iter().zip(got).enumerate() {
                compare_value(wanted, got, &format!("{path}.{index}"), failures);
            }
        }
        (Value::Number(wanted), Value::Number(got)) => {
            if !numbers_equal(wanted, got) {
                failures.push(format!("{path}: got {got}, want {wanted}"));
            }
        }
        (wanted, got) if wanted != got => {
            failures.push(format!("{path}: got {got}, want {wanted}"));
        }
        _ => {}
    }
}

fn numbers_equal(expected: &serde_json::Number, actual: &serde_json::Number) -> bool {
    if let (Some(expected), Some(actual)) = (expected.as_i64(), actual.as_i64()) {
        return expected == actual;
    }
    if let (Some(expected), Some(actual)) = (expected.as_u64(), actual.as_u64()) {
        return expected == actual;
    }
    match (expected.as_f64(), actual.as_f64()) {
        (Some(expected), Some(actual)) => {
            if expected == actual {
                true
            } else {
                let scale = expected.abs().max(actual.abs());
                scale != 0.0 && (expected - actual).abs() / scale <= 1e-9
            }
        }
        _ => false,
    }
}

fn is_default(value: &Value) -> bool {
    match value {
        Value::Null => true,
        Value::Bool(false) => true,
        Value::String(value) if value.is_empty() => true,
        Value::Number(value) => {
            value.as_i64() == Some(0) || value.as_u64() == Some(0) || value.as_f64() == Some(0.0)
        }
        Value::Array(value) => value.is_empty(),
        Value::Object(value) => value.is_empty(),
        _ => false,
    }
}

pub fn lookup<'a>(value: &'a Value, path: &str) -> Option<&'a Value> {
    lookup_parts(value, &path.split('.').collect::<Vec<_>>())
}

fn values_at<'a>(value: &'a Value, path: &str) -> Vec<&'a Value> {
    let parts = path.split('.').collect::<Vec<_>>();
    values_parts(value, &parts)
}

fn values_parts<'a>(value: &'a Value, parts: &[&str]) -> Vec<&'a Value> {
    let Some((head, rest)) = parts.split_first() else {
        return vec![value];
    };
    if *head == "*" {
        match value {
            Value::Array(items) => items
                .iter()
                .flat_map(|item| values_parts(item, rest))
                .collect(),
            Value::Object(items) => items
                .values()
                .flat_map(|item| values_parts(item, rest))
                .collect(),
            _ => Vec::new(),
        }
    } else {
        match value {
            Value::Object(items) => items
                .get(*head)
                .map_or_else(Vec::new, |item| values_parts(item, rest)),
            Value::Array(items) => head
                .parse::<usize>()
                .ok()
                .and_then(|index| items.get(index))
                .map_or_else(Vec::new, |item| values_parts(item, rest)),
            _ => Vec::new(),
        }
    }
}

fn lookup_parts<'a>(value: &'a Value, parts: &[&str]) -> Option<&'a Value> {
    let Some((head, rest)) = parts.split_first() else {
        return Some(value);
    };
    if *head == "*" {
        match value {
            Value::Array(values) => values.iter().find_map(|value| lookup_parts(value, rest)),
            Value::Object(values) => values.values().find_map(|value| lookup_parts(value, rest)),
            _ => None,
        }
    } else {
        let next = match value {
            Value::Object(values) => values.get(*head),
            Value::Array(values) => head
                .parse::<usize>()
                .ok()
                .and_then(|index| values.get(index)),
            _ => None,
        }?;
        lookup_parts(next, rest)
    }
}

/// Fields other than `Instance.id` that carry a runtime instance id: a value's
/// reference to an object and a function's bound object.
pub const RUNTIME_ID_KEYS: [&str; 2] = ["instance_id", "self_id"];

pub fn label_instance_ids(value: &mut Value) {
    let mut labels = HashMap::new();
    label_ids(value, &mut labels);
}

fn label_ids(value: &mut Value, labels: &mut HashMap<i64, String>) {
    match value {
        Value::Object(object) => {
            let is_instance = object.contains_key("type_symbol_id")
                && object.contains_key("feature_values")
                && object.contains_key("id");
            if is_instance {
                if let Some(child) = object.get_mut("id") {
                    label_id(child, labels);
                }
            }
            for key in RUNTIME_ID_KEYS {
                if let Some(child) = object.get_mut(key) {
                    label_id(child, labels);
                }
            }
            for (key, child) in object.iter_mut() {
                let labeled =
                    (is_instance && key == "id") || RUNTIME_ID_KEYS.contains(&key.as_str());
                if !labeled {
                    label_ids(child, labels);
                }
            }
        }
        Value::Array(items) => {
            for item in items {
                label_ids(item, labels);
            }
        }
        _ => {}
    }
}

fn label_id(value: &mut Value, labels: &mut HashMap<i64, String>) {
    if let Some(id) = value.as_i64() {
        let next = format!("@{}", labels.len() + 1);
        let label = labels.entry(id).or_insert(next).clone();
        *value = Value::String(label);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::scenario::Expect;
    use serde_json::json;

    #[test]
    fn relative_real_tolerance_has_both_boundaries() {
        let expect: Expect = serde_json::from_value(json!({"response": 100.0})).unwrap();
        assert!(compare(&expect, &json!(100.0 + 1e-10)).is_empty());
        assert!(!compare(&expect, &json!(100.0 + 1e-6)).is_empty());
    }

    #[test]
    fn named_lists_require_exact_length() {
        let expect: Expect = serde_json::from_value(json!({"response": [1, 2]})).unwrap();
        assert!(!compare(&expect, &json!([1, 2, 3])).is_empty());
        assert!(compare(&expect, &json!([1, 2])).is_empty());
    }

    #[test]
    fn defaults_and_wildcards_are_observable() {
        let actual = json!({"error": "", "elements": [{"id": 1}, {"id": 2}]});
        let expect: Expect = serde_json::from_value(json!({
            "response": {"error": ""},
            "absent": ["missing"],
            "non_empty": ["elements"],
            "contains_all": {"elements.*.id": ["1"]}
        }))
        .unwrap();
        assert!(!compare(&expect, &actual).is_empty());
        let actual = json!({"error": "", "elements": [{"id": "1"}, {"id": "2"}]});
        assert!(compare(&expect, &actual).is_empty());
        let expect_both: Expect = serde_json::from_value(json!({
            "contains_all": {"elements.*.id": ["1", "2"]}
        }))
        .unwrap();
        assert!(compare(&expect_both, &actual).is_empty());
        assert!(!compare(
            &expect_both,
            &json!({"elements": [{"id": "1"}, {"id": "3"}]})
        )
        .is_empty());
        assert_eq!(lookup(&actual, "elements.1.id"), Some(&json!("2")));
    }

    #[test]
    fn instance_ids_are_labelled_consistently() {
        let mut actual = json!({
            "instance": {"id": 9, "type_symbol_id": "T", "feature_values": {}},
            "instances": [{"id": 9, "type_symbol_id": "T", "feature_values": {}},
                          {"id": 12, "type_symbol_id": "T", "feature_values": {}}],
            "value": {"instance_id": 12},
            "function": {"calc_id": "T::f", "self_id": 9}
        });
        label_instance_ids(&mut actual);
        assert_eq!(actual["instance"]["id"], "@1");
        assert_eq!(actual["instances"][0]["id"], "@1");
        assert_eq!(actual["instances"][1]["id"], "@2");
        assert_eq!(actual["value"]["instance_id"], "@2");
        assert_eq!(actual["function"]["self_id"], "@1");
    }

    #[test]
    fn integer_precision_is_not_float_precision() {
        let expect: Expect =
            serde_json::from_value(json!({"response": 9007199254740993i64})).unwrap();
        assert!(compare(&expect, &json!(9007199254740993i64)).is_empty());
        assert!(!compare(&expect, &json!(9007199254740992i64)).is_empty());
    }

    #[test]
    fn statuses_compare_by_canonical_name() {
        assert!(status_matches(Some("NOT_FOUND"), Status::NotFound));
        assert!(!status_matches(Some("not_found"), Status::NotFound));
        assert!(status_matches(None, Status::Ok));
    }
}
